### ISSUE TITLE
Apparent change of fonts when loading any page

### ISSUE THREAD

**Issue by miloyip** [Phlow/feeling-responsive#19]
When I load the pages in https://phlow.github.io/feeling-responsive/ or my blog on Mac, the fonts of all texts changes apparently in a second. It appears both in Chrome and Safari. This especially obvious when loading a page without cache.

I suspect it may due to some loading sequence of javascript or CSS. Or will it be related to the font-family in CSS?

I will also try to do experiments to diagnoses this problem.

**Comment by miloyip** [comment]
After a bit of experiments and research, I found that this is due to web font. And this issue has a specific name called [Flash of Unstyled Text (FOUT)](http://help.typekit.com/customer/portal/articles/6852).

But just setting style of `wf-loading` is not working in my environment. I tried the technique in [here](http://www.position-absolute.com/articles/introduction-to-the-google-webfont-loader-and-how-to-avoid-font-face-text-flickering-with-it/) and find it working.

~~~html
  <script>
    document.getElementsByTagName("html")[0].setAttribute("class","wf-loading");

    WebFontConfig = {
      google: { families: [ 'Lato:400,700,400italic:latin', 'Volkhov::latin' ] }
    };

    (function() {
      var wf = document.createElement('script');
      wf.src = ('https:' == document.location.protocol ? 'https' : 'http') +
                '://ajax.googleapis.com/ajax/libs/webfont/1.5.6/webfont.js';
      wf.type = 'text/javascript';
      wf.async = 'false';
      var s = document.getElementsByTagName('script')[0];
      s.parentNode.insertBefore(wf, s);
    })();
  </script>

  <noscript>
    <link href='http://fonts.googleapis.com/css?family=Lato:400,700,400italic%7cVolkhov' rel='stylesheet' type='text/css'>
  </noscript>

  <style type="text/css">
  .wf-loading body {
      visibility: hidden;
  }
  .wf-inactive body, .wf-active body{
      visibility:visible;
  }
  </style>
~~~

I am not sure if this is a good thing to do. As it will increases the time of completely blank.
I also tried to just make text elements invisible.

~~~html
  <style type="text/css">
  .wf-loading #navigation,
  .wf-loading h1,
  .wf-loading h2,
  .wf-loading p,
  .wf-loading li
  {
      visibility: hidden;
  }
  .wf-inactive #navigation,
  .wf-inactive h1,
  .wf-inactive h2,
  .wf-inactive p,
  .wf-inactive li,
  .wf-active #navigation,
  .wf-active h1,
  .wf-active h2,
  .wf-active p,
  .wf-active li
  {
      visibility: visible;
  }
  </style>
~~~

I have not listed all elements, but the idea is just make texts invisible before the font is loaded.

**Comment by Phlow** [comment]
Hey Milo!

You're right. It's the * Flash of Unstyled Text (FOUT)* problem. To prevent this I embedded the javascript. But I am sorry, it's not done properly by myself. I have to correct this issue. Thanks for pointing out.

It will take some time for me to correct this because I have a lot of projects right now. But I will fix it. If you want to fix it immediately, just use fonts like Arial, Georgia and other preinstalled fonts.

Or **check out [Webfontloader](https://github.com/typekit/webfontloader)** which adresses exactly the flashing font problem.

Thank you for your help.

All the Best, mo.

**Comment by TWiStErRob** [comment]
As far as I understood we have 3 options:
 * async font load with FOUT
 * sync font load without FOUT, but text appearing delayed
 * no custom fonts

I simply replaced the script block with:
```html
<link href='http://fonts.googleapis.com/css?family=Lato:400,700,400italic%7cVolkhov' rel='stylesheet' type='text/css'>
```
it didn't seemed to change the loading time, and it behaves like your `visibility` trick. @miloyip's solution **may be** crossbrowser though. I only tested Chrome.

**Comment by Phlow** [comment]
I want to fix it. I read [Responsive Typography](http://shop.oreilly.com/product/0636920034063.do) but didn't finish my work. I do it after my holidays ;)

**Comment by TWiStErRob** [comment]
Yep, looking forward to that!

**Comment by franfabrizio** [comment]
Just came across this issue today too.  I'm using TWiStErRob's solution for now, but also looking forward to the real fix.  Thanks!

**Comment by rphaedrus** [comment]
I've created https://github.com/Phlow/feeling-responsive/pull/31 to fix this, although I suspect the fix that TWiStErRob used is as good since the google web font loader's advanced features aren't being used.

**Comment by Phlow** [comment]
Haven't tried it yet, but I think it's fixed thanks to @rphaedrus :)

**Comment by TWiStErRob** [comment]
@rphaedrus there's one other big difference between your fix and mine. Though they're both sync, the JS wrapper needs JS (feature enabled and resources for download/parse/execute), while a simple `<link>` is only CSS (loaded anyway by JS too).

You're right though, one extra it gives is the hooks, but I'm not sure how/why they're helpful in general.

Best would be a delayless solution which doesn't load the font first time, but only when it's already cached.

**Comment by Phlow** [comment]
As soon as I have some time in Gemany, I will look into this. I got the idea by a good book on responsive typography on how to prevent the flash, so that the visitors can read the content immidiately. when the font is loaded, it will replace the old one. that's what the javascript is for. I hope I fix it in august. Please be patient.

**Comment by TWiStErRob** [comment]
No worries, I remember, enjoy your stay! I was just responding to rphaedrus's comment on #31. Thanks for the JS hook clarification.

**Comment by rphaedrus** [comment]
The one I posted includes the <link > tag wrapped in <noscript>, so we should be good for the non-JS situation although you're right about the resources.

Loading the page the first time with default fonts while caching the new fonts behind the scenes to be loaded the second time is a really interesting idea.  I've not encountered that before.  I'm looking forward to seeing how it feels.

**Comment by Phlow** [comment]
This is for people using slow connections... Normally, you don't see it.

**Comment by TWiStErRob** [comment]
@Phlow I saw FOUT every time I pressed F5 in my dev env. Even when the site wasn't recompiled.

@rphaedrus https://css-tricks.com/preventing-the-performance-hit-from-custom-fonts/; some article also claimed The Guardian used [webfontjson](https://github.com/ahume/webfontjson).
I also found [localFont](https://github.com/jaicab/localFont), but it lost a little credibility when I clicked on "the article" link and saw a FOUT, only the first time though.
